palace/models/waveportoperator.cpp

```diff
@@ -321,290 +474,315 @@ class BdrHVectorCoefficient : public mfem::VectorCoefficient
       slepc->SetType(slepc::SlepcEigenSolver::Type::KRYLOVSCHUR);
+      slepc->SetProblemType(slepc::SlepcEigenSolver::ProblemType::GEN_NON_HERMITIAN);
+      eigen = std::move(slepc);
 #endif
     }
     constexpr double tol = 1.0e-6;
-    eigen->SetLinearSolver(*ksp);
-    eigen->SetWhichEigenpairs(EigenSolverBase::WhichType::LARGEST_MAGNITUDE);
     eigen->SetNumModes(mode_idx, std::max(2 * mode_idx + 1, 5));
     eigen->SetTol(tol);
+    eigen->SetWhichEigenpairs(EigenvalueSolver::WhichType::LARGEST_MAGNITUDE);
+    eigen->SetLinearSolver(*ksp);
   }
-}
 
-void WavePortData::GetTrueDofs(const mfem::Array<int> &dbc_marker,
-                               mfem::ParFiniteElementSpace &nd_fespace,
-                               mfem::ParFiniteElementSpace &h1_fespace,
-                               mfem::Array<int> &nd_tdof_list,
-                               mfem::Array<int> &h1_tdof_list)
-{
-  // Ensures no duplicates in the attribute list for this port index (this would imply a
-  // mistake in the configuration file). We can, however, have multiple unique ports with
-  // shared boundary attributes.
-  nd_fespace.GetEssentialTrueDofs(attr_marker, nd_tdof_list);
-  h1_fespace.GetEssentialTrueDofs(attr_marker, h1_tdof_list);
-  int nd_tdofs = nd_tdof_list.Size();
-  int h1_tdofs = h1_tdof_list.Size();
-
-  // Mark all ND and H1 dofs on the port, then unmark PEC boundaries.
-  mfem::Array<int> nd_tdof_marker(nd_fespace.GetTrueVSize()),
-      h1_tdof_marker(h1_fespace.GetTrueVSize()), nd_dbc_tdof_list, h1_dbc_tdof_list;
-  nd_tdof_marker = 0;
-  h1_tdof_marker = 0;
-  nd_fespace.GetEssentialTrueDofs(dbc_marker, nd_dbc_tdof_list);
-  h1_fespace.GetEssentialTrueDofs(dbc_marker, h1_dbc_tdof_list);
-  for (auto tdof : nd_tdof_list)
-  {
-    nd_tdof_marker[tdof] = 1;
-  }
-  for (auto tdof : nd_dbc_tdof_list)
-  {
-    nd_tdof_marker[tdof] = 0;
-  }
-  for (auto tdof : h1_tdof_list)
-  {
-    h1_tdof_marker[tdof] = 1;
-  }
-  for (auto tdof : h1_dbc_tdof_list)
+  // Coefficients store references to kₙ, ω so they are updated implicitly at each new
+  // solve. Also, μ⁻¹ is persistent, so no copy is OK.
+  kn0 = 0.0;
+  omega0 = 0.0;
+  port_nxH0r_func =
+      std::make_unique<BdrSubmeshHVectorCoefficient<true>>(*port_E0t, *port_E0n, mat_op);
+  port_nxH0i_func =
+      std::make_unique<BdrSubmeshHVectorCoefficient<false>>(*port_E0t, *port_E0n, mat_op);
+  port_sr = std::make_unique<mfem::LinearForm>(port_nd_fespace.get());
+  port_si = std::make_unique<mfem::LinearForm>(port_nd_fespace.get());
+  port_sr->AddDomainIntegrator(new VectorFEDomainLFIntegrator(*port_nxH0r_func));
+  port_si->AddDomainIntegrator(new VectorFEDomainLFIntegrator(*port_nxH0i_func));
+  port_sr->UseFastAssembly(false);
+  port_si->UseFastAssembly(false);
+
+  // Configure port mode sign convention: 1ᵀ Re{-n x H} >= 0 on the "upper-right quadrant"
+  // of the wave port boundary, in order to deal with symmetry effectively.
```

Comment:
Wave ports  actually still work currently, no matter if the bounding box is actually axis aligned for the port, since this is just used to give a consistent +/- orientation for the mode.